crates/adapters/src/integrated/postgres/output.rs

```diff
@@ -0,0 +1,299 @@
+            client,
+            transaction: None,
+            insert,
+            delete,
+            keys,
+            _pin: PhantomPinned,
+        };
+
+        let _guard = out.span();
+
+        Ok(out)
+    }
+
+    fn transaction(&mut self) -> AnyResult<&mut postgres::Transaction<'static>> {
+        self.transaction.as_mut().ok_or(anyhow!("postgres: unreachable: attempting to perform a transaction that hasn't been created yet"))
+    }
+
+    pub fn insert(&mut self, value: &str) -> AnyResult<()> {
+        if value.len() <= 2 {
+            return Ok(());
+        }
+
+        let ins = self.insert.clone();
+        self.transaction()?.execute(&ins, &[&value])?;
+        Ok(())
+    }
+
+    pub fn delete(&mut self, value: &str) -> AnyResult<()> {
+        if value.len() <= 2 {
+            return Ok(());
+        }
+
+        let del = self.delete.clone();
+        self.transaction()?.execute(&del, &[&value])?;
+        Ok(())
+    }
+
+    pub fn span(&self) -> EnteredSpan {
+        info_span!(
+            "postgres_output",
+            ft = false,
+            id = self.endpoint_id,
+            name = self.endpoint_name,
+            pg_table = self.table,
+        )
+        .entered()
+    }
+}
+
+impl OutputConsumer for PostgresOutputEndpoint {
+    fn max_buffer_size_bytes(&self) -> usize {
+        usize::MAX
+    }
+
+    fn batch_start(&mut self, _: Step) {
+        let transaction: postgres::Transaction<'static> = unsafe {
+            std::mem::transmute(
+                self.client
+                    .transaction()
+                    .expect("postgres: failed to start transaction"),
+            )
+        };
+
+        self.transaction = Some(transaction);
+        self.start = Instant::now();
+    }
+
+    fn push_buffer(&mut self, _: &[u8], _: usize) {
+        unreachable!()
+    }
+
+    fn push_key(
+        &mut self,
+        _: Option<&[u8]>,
+        _: Option<&[u8]>,
+        _: &[(&str, Option<&[u8]>)],
+        num_records: usize,
+    ) {
+        unreachable!()
+    }
+
+    fn batch_end(&mut self) {
+        let transaction: postgres::Transaction<'static> = self.transaction.take().expect(
+            "postgres: unreachable: attempted to commit a transaction that hasn't been started",
+        );
+        transaction
+            .commit()
+            .expect("postgres: failed to commit transaction");
+        let elapsed = self.start.elapsed();
+
+        tracing::info!("transaction took: {elapsed:?}",);
+    }
+}
+
+impl Encoder for PostgresOutputEndpoint {
+    fn consumer(&mut self) -> &mut dyn OutputConsumer {
+        self
+    }
+
+    fn encode(&mut self, batch: &dyn SerBatchReader) -> anyhow::Result<()> {
+        let mut insert_buffer = "[".to_owned();
+        let mut delete_buffer = "[".to_owned();
+
+        let mut cursor =
+            CursorWithPolarity::new(batch.cursor(RecordFormat::Json(Default::default()))?);
+
+        while cursor.key_valid() {
```

Comment:
This code assumes one value per key, which is not the case with `index`. See Avro output connector for an example of handling indexed outptus (or ping me to go over it).